### PR TITLE
Switch from enums to object literals to reduce the bundle size

### PR DESCRIPTION
**Pull request by konstantin-msft**
- Switch from enums to object literals to reduce the bundle size.
- Add eslint rule.

### CODE REVIEW

**Comment by tnorling** [comment]
I'm not sure that we can turn every `enum` into a `const enum`. Works for things that are completely internal but how does this work for things that are exported from the lib?

File: change/@azure-msal-browser-0869b94e-bc69-484e-96e8-c9dd124dca5b.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "patch",
```
Comment (hectormmg) [inline change/@azure-msal-browser-0869b94e-bc69-484e-96e8-c9dd124dca5b.json]:
```suggestion
  "type": "prerelease",
```

File: change/@azure-msal-browser-0869b94e-bc69-484e-96e8-c9dd124dca5b.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+  "comment": "Switch to const enums to reduce the bundle size #6056",
+  "packageName": "@azure/msal-browser",
+  "email": "kshabelko@microsoft.com",
+  "dependentChangeType": "patch"
```
Comment (hectormmg) [inline change/@azure-msal-browser-0869b94e-bc69-484e-96e8-c9dd124dca5b.json]:
```suggestion
  "dependentChangeType": "prerelease"
```

File: change/@azure-msal-common-089b1ea1-cf4f-4f55-b3d6-2877c26e652b.json [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "patch",
```
Comment (hectormmg) [inline change/@azure-msal-common-089b1ea1-cf4f-4f55-b3d6-2877c26e652b.json]:
```suggestion
  "type": "prerelease",
```

File: change/@azure-msal-common-089b1ea1-cf4f-4f55-b3d6-2877c26e652b.json [hunk 4]
```diff
@@ -0,0 +1,7 @@
...
+  "comment": "Switch to const enums to reduce the bundle size #6056",
+  "packageName": "@azure/msal-common",
+  "email": "kshabelko@microsoft.com",
+  "dependentChangeType": "patch"
```
Comment (hectormmg) [inline change/@azure-msal-common-089b1ea1-cf4f-4f55-b3d6-2877c26e652b.json]:
```suggestion
  "dependentChangeType": "prerelease"
```

**Review by hectormmg — APPROVED** [review]
LGTM, out of curiosity, what was the bundle size reduction?

**Comment by hectormmg** [comment]
> I'm not sure that we can turn every `enum` into a `const enum`. Works for things that are completely internal but how does this work for things that are exported from the lib?

Great point, I agree this change has to be limited to internal enums only. This also applies to enums exported from common that are used in browser and node.

File: change/@azure-msal-browser-0869b94e-bc69-484e-96e8-c9dd124dca5b.json (hunk 1 above)
Comment (konstantin-msft) [reply change/@azure-msal-browser-0869b94e-bc69-484e-96e8-c9dd124dca5b.json]: fixed

File: change/@azure-msal-browser-0869b94e-bc69-484e-96e8-c9dd124dca5b.json (hunk 2 above)
Comment (konstantin-msft) [reply change/@azure-msal-browser-0869b94e-bc69-484e-96e8-c9dd124dca5b.json]: fixed

File: change/@azure-msal-common-089b1ea1-cf4f-4f55-b3d6-2877c26e652b.json (hunk 3 above)
Comment (konstantin-msft) [reply change/@azure-msal-common-089b1ea1-cf4f-4f55-b3d6-2877c26e652b.json]: fixed

File: change/@azure-msal-common-089b1ea1-cf4f-4f55-b3d6-2877c26e652b.json (hunk 4 above)
Comment (konstantin-msft) [reply change/@azure-msal-common-089b1ea1-cf4f-4f55-b3d6-2877c26e652b.json]: fixed

**Comment by konstantin-msft** [comment]
> LGTM, out of curiosity, what was the bundle size reduction?

It's ~11Kb.

**Comment by tnorling** [comment]
> > I'm not sure that we can turn every `enum` into a `const enum`. Works for things that are completely internal but how does this work for things that are exported from the lib?

There's a suggestion for handling precisely this problem in the TypeScript docs: https://www.typescriptlang.org/docs/handbook/enums.html#const-enum-pitfalls

**Comment by konstantin-msft** [comment]
> > > I'm not sure that we can turn every `enum` into a `const enum`. Works for things that are completely internal but how does this work for things that are exported from the lib?
> 
> There's a suggestion for handling precisely this problem in the TypeScript docs: https://www.typescriptlang.org/docs/handbook/enums.html#const-enum-pitfalls

Thanks! Came across the same solution just an hour ago. I will give it a try to check if we can get rid of enums completely.

**Review by tnorling — APPROVED** [review]

**Comment by ghost** [comment]
:tada:`@azure/msal-common@v14.0.0-beta.0` has been released which incorporates this pull request.:tada:

We recommend upgrading to the latest version of `@azure/msal-browser` or `@azure/msal-node` to take advantage of this change.

Handy links:
* [Release Notes](https://github.com/AzureAD/microsoft-authentication-library-for-js/releases/tag/msal-common-v14.0.0-beta.0)
* [Official NPM Release](https://www.npmjs.com/package/@azure/msal-common/v/14.0.0)

**Pull request merged**